src/go/rpk/pkg/cli/cmd/debug/bundle.go

```diff
@@ -0,0 +1,842 @@
+// Copyright 2021 Vectorized, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+//go:build linux
+// +build linux
+
+package debug
+
+import (
+	"archive/zip"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"os/exec"
+	"os/user"
+	"path/filepath"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/Shopify/sarama"
+	"github.com/beevik/ntp"
+	"github.com/docker/go-units"
+	"github.com/hashicorp/go-multierror"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/afero"
+	"github.com/spf13/cobra"
+	"github.com/vectorizedio/redpanda/src/go/rpk/pkg/api/admin"
+	"github.com/vectorizedio/redpanda/src/go/rpk/pkg/cli/cmd/common"
+	"github.com/vectorizedio/redpanda/src/go/rpk/pkg/config"
+	"github.com/vectorizedio/redpanda/src/go/rpk/pkg/system"
+	"github.com/vectorizedio/redpanda/src/go/rpk/pkg/system/syslog"
+	"gopkg.in/yaml.v2"
+)
+
+// Use the same date specs as journalctl (see `man journalctl`).
+const timeHelpText = `(journalctl date format, e.g. YYYY-MM-DD)`
+
+type step func() error
+
+type stepParams struct {
+	fs      afero.Fs
+	m       *sync.Mutex
+	w       *zip.Writer
+	timeout time.Duration
+}
+
+type limitedWriter struct {
+	w          io.Writer
+	limitBytes int
+	accumBytes int
+}
+
+func (l *limitedWriter) Write(p []byte) (int, error) {
+	n, err := l.w.Write(p)
+	if err != nil {
+		return n, err
+	}
+
+	l.accumBytes += n
+
+	if l.accumBytes >= l.limitBytes {
```

Comment:
If we want to stop at the limit, we should check `l.accumBytes + len(p) >= l.limitBytes` before the write (and truncate p accordingly, or not perform the write).

Reply:
Sounds good. I intentionally left it like this (not strict) because each write is a block long, which on many machines is 4kB. But I agree, let's make it strict.

Reply:
Fixed.

Reply:
I think this will fail if the limit is 100 and we write exactly 100 bytes, instead we may want to change the initial conditional to:
```go
if l.accumBytes+len(p) > l.limitBytes {
  p = ...
  limitReached = true
}
```
and then return err here `if limitReached`

Reply:
Done.